### PR TITLE
chore: addition of new GQL subscription for `UserCollection` duplication

### PR DESCRIPTION
**Pull request by balub**
Closes HSB-476

### What's changed

In this PR we add a new GQL subscription `userCollectionDuplicated` which returns the entire UserCollection tree when subscribed to.

### Notes to reviewers

Nil

### CODE REVIEW

File: packages/hoppscotch-backend/src/user-collection/user-collection.service.ts [hunk 1]
```diff
@@ -1035,6 +1036,7 @@ export class UserCollectionService {
     userID: string,
     destCollectionID: string | null,
     reqType: DBReqType,
+    isCollectionDuplication = true,
```
Comment (jamesgeorge007) [inline packages/hoppscotch-backend/src/user-collection/user-collection.service.ts:1039]:
```suggestion
    isCollectionDuplication = false,
```

File: packages/hoppscotch-backend/src/user-collection/user-collection.service.ts [hunk 2]
```diff
@@ -1087,6 +1089,17 @@ export class UserCollectionService {
...
+      if (!E.isLeft(collectionData))
+        this.pubsub.publish(
+          `user_coll/${userID}/duplicated`,
+          collectionData.right,
+        );
+    }
+
     userCollections.forEach((collection) =>
       this.pubsub.publish(`user_coll/${userID}/created`, this.cast(collection)),
     );
```
Comment (jamesgeorge007) [inline packages/hoppscotch-backend/src/user-collection/user-collection.service.ts:1105]: We require only the `.../duplicated` topic to be published during a duplicate action. Also, shouldn't the check be the other way around where the `.../duplicated` topic is published if `isCollectionDuplication` is set to `true`?

```suggestion
    if (isCollectionDuplication) {
      const collectionData = await this.fetchCollectionData(
        userCollections[0].id,
      );

      if (E.isRight(collectionData))
        this.pubsub.publish(
          `user_coll/${userID}/duplicated`,
          collectionData.right,
        );
    } else {
      userCollections.forEach((collection) =>
        this.pubsub.publish(
          `user_coll/${userID}/created`,
          this.cast(collection),
        ),
      );
    }
```

File: packages/hoppscotch-backend/src/user-collection/user-collection.service.ts [hunk 3]
```diff
@@ -1182,9 +1195,55 @@ export class UserCollectionService {
...
     return E.right(true);
   }
+
+  /**
+   * Generates a JSON containing all the contents of a collection
+   *
+   * @param collection Collection whose details we want to fetch
+   * @returns A JSON string containing all the contents of a collection
+   */
+  private async fetchCollectionData(collectionID: string) {
```
Comment (jamesgeorge007) [inline packages/hoppscotch-backend/src/user-collection/user-collection.service.ts:1211]: Noticed the return type wasn't getting inferred correctly and might want to specify it.

```suggestion
  private async fetchCollectionData(
    collectionID: string,
  ): Promise<E.Left<string> | E.Right<UserCollectionDuplicatedData>> {
```

File: packages/hoppscotch-backend/src/user-collection/user-collection.service.ts [hunk 4]
```diff
@@ -1182,9 +1195,55 @@ export class UserCollectionService {
       userID,
       collection.right.parentID,
       reqType,
+      false,
```
Comment (jamesgeorge007) [inline packages/hoppscotch-backend/src/user-collection/user-collection.service.ts:1198]:
```suggestion
      true,
```

File: packages/hoppscotch-backend/src/user-collection/user-collection.resolver.ts [hunk 5]
```diff
@@ -36,6 +37,7 @@ import { ReqType } from 'src/types/RequestTypes';
 import * as E from 'fp-ts/Either';
 import { GqlThrottlerGuard } from 'src/guards/gql-throttler.guard';
 import { SkipThrottle } from '@nestjs/throttler';
+import { any } from 'jest-mock-extended';
```
Comment (mirarifhasan) [inline packages/hoppscotch-backend/src/user-collection/user-collection.resolver.ts:40]: Remove unused imports

**Review by mirarifhasan — CHANGES_REQUESTED** [review]

File: docker-compose.yml
Comment (jamesgeorge007) [inline docker-compose.yml:1]: Please be sure to revert these changes before merging.

**Review by jamesgeorge007 — APPROVED** [review]

**Review by mirarifhasan — APPROVED** [review]
LGTM

**Review by jamesgeorge007 — APPROVED** [review]

**Pull request merged**